nncf/torch/kd_loss.py

```diff
@@ -0,0 +1,148 @@
+from nncf.torch.compression_method_api import PTCompressionAlgorithmController
+from nncf.api.compression import CompressionLevel, CompressionLoss, CompressionScheduler
+from nncf.torch.algo_selector import COMPRESSION_ALGORITHMS
+from nncf.torch.compression_method_api import PTCompressionLoss
+
+KD_MODULE_NAME = 'KD_FP32_MODULE'
+
+
+class KDLossCalculator(PTCompressionLoss):
+    def __init__(self, target_model: NNCFNetwork, original_model, scale, type):
+        super().__init__()
+        self._target_model = target_model
+        self.original_model = original_model
+        self.original_model.train()
+        self.scale = scale
+        if type == 'softmax':
+            def kdloss_fn(ref_outputs, compressed_model_outputs):
+                return -(nn.functional.log_softmax(compressed_model_outputs, dim=1) *
+                         nn.functional.softmax(ref_outputs, dim=1)).mean() * (compressed_model_outputs.shape[1])
+        elif type == 'mse':
+            def kdloss_fn(ref_outputs, compressed_model_outputs):
+                mse = torch.nn.MSELoss()
+                return mse(ref_outputs, compressed_model_outputs)
+        else:
+            raise ValueError('Choose between mse/softmax options for Knowledge Distillation')
+        self.kdloss_fn = kdloss_fn
+
+    def calculate_kd_loss(self, compressed_model_outputs, orig_model_outputs):
+        """
+        Calculates KD Loss value from compressed_model_outputs and orig_model_outputs. First uses
+        nested_object_paths_generator to unpack input containers and numerate contents inside them.
+        Than checks compressed_model_outputs unpacked container for loss tensors (requires_grad=True)
+        and maps extracted structure of loss tensors to orig_model_outputs.
+        Finally computes KD loss with extracted loss tensors.
+
+        :param compressed_model_outputs: Output tensors of compressed model can be any type of container with
+            deterministic detour
+        :param orig_model_outputs: Output tensors of original model (used for distillation) can be any type of 
+            container with deterministic detour.
+        :return: KDLoss value
+        """
+
+        compressed_model_outputs_struct = []
+        orig_model_outputs_struct = []
+        OperatorInput.nested_object_paths_generator([compressed_model_outputs], compressed_model_outputs_struct)
+        OperatorInput.nested_object_paths_generator([orig_model_outputs], orig_model_outputs_struct)
+        compressed_model_loss_nested_obj_paths = list(filter(lambda x: self._is_loss(x.getter()),
+                                                             compressed_model_outputs_struct))
+        compressed_model_loss_outputs = list(map(lambda x: x.getter(), compressed_model_loss_nested_obj_paths))
+        orig_model_loss_outputs = list(map(lambda x: x.getter(), filter(
+            partial(self.match_func, to_match_with=compressed_model_loss_nested_obj_paths), orig_model_outputs_struct)))
```

Comment:
Don't think you need to make it so general with partial and static function.
Why not just defining specific lambda function with `compressed_model_loss_nested_obj_paths` instead of `to_match_with`? It would be more readable.